apis/r/R/SOMADataFrame.R

```diff
@@ -225,7 +218,118 @@ SOMADataFrame <- R6::R6Class(
                      loglevel = log_level)
 
       TableReadIter$new(sr)
+    },
+
+    #' @description Update (lifecycle: experimental)
+    #' @details
+    #' Update the existing `SOMADataFrame` to add or remove columns based on the
+    #' input:
+    #' - columns present in the current the `SOMADataFrame` but absent from the
+    #'   new `values` will be dropped
+    #' - columns absent in current `SOMADataFrame` but present in the new
+    #'   `values` will be added
+    #' - any columns present in both will be left alone, with the exception that
+    #'   if `values` has a different type for the column, the entire update
+    #'   will because attribute types cannot be changed
+    #'
+    #' @param values A `data.frame`, [`arrow::Table`], or
+    #' [`arrow::RecordBatch`].
+    #' @param row_index_name An optional scalar character. If provided, and if
+    #' the `values` argument is a `data.frame` with row names, then the row
+    #' names will be extracted and added as a new column to the `data.frame`
+    #' prior to performing the update. The name of this new column will be set
+    #' to the value specified by `row_index_name`.
+
+    update = function(values, row_index_name = NULL) {
+      private$check_open_for_write()
+      stopifnot(
+        "'values' must be a data.frame, Arrow Table or RecordBatch" =
+          is.data.frame(values) || is_arrow_table(values) || is_arrow_record_batch(values)
+      )
+
+      if (is.data.frame(values)) {
+        if (!is.null(row_index_name)) {
+          stopifnot(
+            "'row_index_name' must be NULL or a scalar character vector" =
+              is_scalar_character(row_index_name),
+            "'row_index_name' conflicts with an existing column name" =
+              !row_index_name %in% colnames(values)
+          )
+          values[[row_index_name]] <- rownames(values)
+        }
+        values <- arrow::as_arrow_table(values)
+      }
+
+      # Retrieve existing soma_joinids from array to:
+      # - validate number of rows in values matches number of rows in array
+      # - add original soma_joinids to values if not present
```

Comment:
You might consider adding a comment to the effect that this suffices _only_ until the day we start implementing row-deletes

Reply:
IDK what if there were 10, 0-9, and they deleted number 2, so the old obs has 0-1 and 3-9. But the new obs has 0-4 and 6-9. (Artificial example I know, and we can discuss in the future since it's a moot point for this PR.)

Reply:
I'm pulling the old values from the array, not generating new ones. So if joinid 2 had been deleted at some point, I'd retrieve values [0,1,3–9] from the array and insert that vector into a new `soma_joinid` column in values.